fggs/semirings.py

```diff
@@ -78,13 +95,16 @@ def solve(self, a: torch.Tensor, b: torch.Tensor) -> torch.Tensor:
         """
         a = a.clone()
         x = b.clone()
+        block_sizes = get_block_sizes(self.block_size, a, a.shape)
         for k in range(a.shape[0]):
-            a[:,k]    = self.mul(a[:,k], self.star(a[k,k]))
-            a[:,k+1:] = self.add(a[:,k+1:], self.mul(a[:,k,None], a[k,k+1:]))
+            a[:,k]           = self.mul(a[:,k], self.star(a[k,k]))
+            for rows, cols in equation.block_sizes_to_indexes((a.shape[0], a.shape[1]-k-1), block_sizes):
+                cols = slice(cols.start+k+1, cols.stop+k+1, cols.step)
+                a[rows,cols] = self.add(a[rows,cols], self.mul(a[rows,k,None], a[k,cols]))
```

Comment:
My `ulimit -v` is 64GB but I had to lower `max_cpu_bytes` to 4GB in order for this `mul` to not run out of memory. I guess it's because there are other things in memory already. It would be nice for `get_available_bytes` to take current memory use into account.